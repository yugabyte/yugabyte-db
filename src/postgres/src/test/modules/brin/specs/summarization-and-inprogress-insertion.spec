# This test verifies that values inserted in transactions still in progress
# are considered during concurrent range summarization (either using the
# brin_summarize_new_values function or regular VACUUM).

setup
{
    CREATE TABLE brin_iso (
        value int
    ) WITH (fillfactor=10);
    CREATE INDEX brinidx ON brin_iso USING brin (value) WITH (pages_per_range=1);
    -- this fills the first page
    INSERT INTO brin_iso VALUES (NULL);
    DO $$
    DECLARE curtid tid;
    BEGIN
      LOOP
        INSERT INTO brin_iso VALUES (1) RETURNING ctid INTO curtid;
        EXIT WHEN curtid > tid '(1, 0)';
      END LOOP;
    END;
    $$;
    CREATE EXTENSION IF NOT EXISTS pageinspect;
}

teardown
{
    DROP TABLE brin_iso;
}

session "s1"
step "s1b"		{ BEGIN ISOLATION LEVEL REPEATABLE READ; }
step "s1i"		{ INSERT INTO brin_iso VALUES (1000); }
step "s1c"		{ COMMIT; }

session "s2"
step "s2b"		{ BEGIN ISOLATION LEVEL REPEATABLE READ; SELECT 1; }
step "s2summ"	{ SELECT brin_summarize_new_values('brinidx'::regclass); }
step "s2c"		{ COMMIT; }

step "s2vacuum"	{ VACUUM brin_iso; }

step "s2check"	{ SELECT * FROM brin_page_items(get_raw_page('brinidx', 2), 'brinidx'::regclass); }

permutation "s2check" "s1b" "s2b" "s1i" "s2summ" "s1c" "s2c" "s2check"
permutation "s2check" "s1b" "s1i" "s2vacuum" "s1c" "s2check"
