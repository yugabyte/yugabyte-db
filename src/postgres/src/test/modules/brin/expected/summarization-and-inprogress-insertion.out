Parsed test spec with 2 sessions

starting permutation: s2check s1b s2b s1i s2summ s1c s2c s2check
step s2check: SELECT * FROM brin_page_items(get_raw_page('brinidx', 2), 'brinidx'::regclass);
itemoffset|blknum|attnum|allnulls|hasnulls|placeholder|value   
----------+------+------+--------+--------+-----------+--------
         1|     0|     1|f       |f       |f          |{1 .. 1}
(1 row)

step s1b: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2b: BEGIN ISOLATION LEVEL REPEATABLE READ; SELECT 1;
?column?
--------
       1
(1 row)

step s1i: INSERT INTO brin_iso VALUES (1000);
step s2summ: SELECT brin_summarize_new_values('brinidx'::regclass);
brin_summarize_new_values
-------------------------
                        1
(1 row)

step s1c: COMMIT;
step s2c: COMMIT;
step s2check: SELECT * FROM brin_page_items(get_raw_page('brinidx', 2), 'brinidx'::regclass);
itemoffset|blknum|attnum|allnulls|hasnulls|placeholder|value      
----------+------+------+--------+--------+-----------+-----------
         1|     0|     1|f       |f       |f          |{1 .. 1}   
         2|     1|     1|f       |f       |f          |{1 .. 1000}
(2 rows)


starting permutation: s2check s1b s1i s2vacuum s1c s2check
step s2check: SELECT * FROM brin_page_items(get_raw_page('brinidx', 2), 'brinidx'::regclass);
itemoffset|blknum|attnum|allnulls|hasnulls|placeholder|value   
----------+------+------+--------+--------+-----------+--------
         1|     0|     1|f       |f       |f          |{1 .. 1}
(1 row)

step s1b: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s1i: INSERT INTO brin_iso VALUES (1000);
step s2vacuum: VACUUM brin_iso;
step s1c: COMMIT;
step s2check: SELECT * FROM brin_page_items(get_raw_page('brinidx', 2), 'brinidx'::regclass);
itemoffset|blknum|attnum|allnulls|hasnulls|placeholder|value      
----------+------+------+--------+--------+-----------+-----------
         1|     0|     1|f       |f       |f          |{1 .. 1}   
         2|     1|     1|f       |f       |f          |{1 .. 1000}
(2 rows)

